### PR TITLE
Bug 1465309 ubuntu 16 puppet at boot setup

### PR DESCRIPTION
**Pull request by dragoscrisan**
The atboot setup for puppet was not working under systemd.

- created systemd service file
- enabled puppet service under Ubuntu 16.04 machines

### CODE REVIEW

File: modules/puppet/files/puppet.service [hunk 1]
```diff
@@ -0,0 +1,10 @@
+[Unit]
+Description=Puppet agent
+
+[Service]
+Type=oneshot
+EnvironmentFile=-/etc/default/puppet
+ExecStart=/etc/puppet/init start
+
+[Install]
+WantedBy=multi-user.target
```
Comment (dividehex) [inline modules/puppet/files/puppet.service]: Don't forget to put a newline at the end of files

File: modules/puppet/manifests/atboot.pp [hunk 2]
```diff
@@ -110,26 +110,60 @@
...
+                    # reload systemd daemon
+                    exec {
+                        'reload systemd':
+                            command => '/bin/systemctl daemon-reload';
+                    }
+                    # enable the service but not start it
+                    service {
+                        'puppet':
+                            enable  => true,
+                            require => File['/lib/systemd/system/puppet.service'];
```
Comment (dividehex) [inline modules/puppet/manifests/atboot.pp]: You probably need to explicitly set the provider as systemd.  We are still dealing with ancient puppet here that can't detect ubuntu 16.04 is a systemd init system.

**Comment by dragoscrisan** [comment]
Tested code on t-linux64-ms-280:
```

[root@t-linux64-ms-280 ~]# puppet agent -t --environment=dcrisan --server=releng-puppet2.srv.releng.scl3.mozilla.com
 Info: Retrieving pluginfacts
 Info: Retrieving plugin
 Info: Loading facts
 Info: Caching catalog for t-linux64-ms-280.test.releng.mdc1.mozilla.com
 Info: Applying configuration version 'heads/Bug_1465309_Ubuntu_16_puppet_at-boot_setup-dirty'
 Notice: /Stage[main]/Disableservices::Common/Service[lightdm]/enable: enable changed 'true' to 'false'
 Notice: /Stage[main]/Puppet::Atboot/File[/etc/puppet/puppetmasters.txt]/content: 
 --- /etc/puppet/puppetmasters.txt	2018-06-26 04:58:48.839994674 -0700
 +++ /tmp/puppet-file20180626-11476-1rmzr4w	2018-06-26 05:10:46.260001881 -0700
 @@ -1,8 +1,8 @@
  releng-puppet1.srv.releng.mdc1.mozilla.com
  releng-puppet2.srv.releng.mdc1.mozilla.com
 -releng-puppet1.srv.releng.mdc2.mozilla.com
  releng-puppet1.srv.releng.use1.mozilla.com
 -releng-puppet2.srv.releng.mdc2.mozilla.com
  releng-puppet2.srv.releng.scl3.mozilla.com
 +releng-puppet2.srv.releng.mdc2.mozilla.com
  releng-puppet1.srv.releng.usw2.mozilla.com
 +releng-puppet1.srv.releng.mdc2.mozilla.com
  releng-puppet1.srv.releng.scl3.mozilla.com
 
 Notice: /Stage[main]/Puppet::Atboot/File[/etc/puppet/puppetmasters.txt]/content: content changed '{md5}b7a0602da5b4ee1cfda58b5fd4ddbec3' to '{md5}881293d11b9858444ed290540dcec75f'
 Notice: /Stage[main]/Disableservices::Common/Service[acpid]/ensure: ensure changed 'running' to 'stopped'
 Notice: /Stage[main]/Mig::Agent::Daemon/Exec[kill mig]/returns: executed successfully
 Info: /Stage[main]/Mig::Agent::Daemon/Exec[kill mig]: Scheduling refresh of Service[mig-agent]
 Notice: /Stage[main]/Mig::Agent::Daemon/Service[mig-agent]: Triggered 'refresh' from 1 events
 Notice: /Stage[main]/Puppet::Atboot/File[/lib/systemd/system/puppet.service]/content: 
 --- /lib/systemd/system/puppet.service	2018-06-25 06:08:30.635496092 -0700
 +++ /tmp/puppet-file20180626-11476-4ve5q9	2018-06-26 05:10:50.236001921 -0700
 @@ -7,4 +7,4 @@
  ExecStart=/etc/puppet/init start
  
  [Install]
 -WantedBy=multi-user.target
 \ No newline at end of file
 +WantedBy=multi-user.target
 
 Notice: /Stage[main]/Puppet::Atboot/File[/lib/systemd/system/puppet.service]/content: content changed '{md5}4b5223b2365ea343a1c7170bad3e4a11' to '{md5}f31dea6170efe15870a2202ad958f6ac'
 Info: /Stage[main]/Puppet::Atboot/File[/lib/systemd/system/puppet.service]: Scheduling refresh of Exec[reload systemd]
 Notice: /Stage[main]/Puppet::Atboot/Exec[reload systemd]/returns: executed successfully
 Notice: /Stage[main]/Puppet::Atboot/Exec[reload systemd]: Triggered 'refresh' from 1 events
 Notice: Finished catalog run in 10.37 seconds

```

**Comment by dragoscrisan** [comment]
applying on an already set-up ubuntu worker (tested on t-linux64-ms-572):
```

[root@t-linux64-ms-572 ~]# puppet agent --test --environment=dcrisan --server=releng-puppet2.srv.releng.scl3.mozilla.com
Info: Retrieving pluginfacts
Info: Retrieving plugin
Info: Loading facts
Info: Caching catalog for t-linux64-ms-572.test.releng.mdc2.mozilla.com
Info: Applying configuration version 'heads/Bug_1465309_Ubuntu_16_puppet_at-boot_setup-dirty'
Notice: /Stage[main]/Packages::Linux_desktop/Package[ubuntu-desktop]/ensure: ensure changed 'purged' to 'latest'
Notice: /Stage[main]/Disableservices::Common/Service[lightdm]/enable: enable changed 'true' to 'false'
Notice: /Stage[main]/Disableservices::Common/Service[acpid]/ensure: ensure changed 'running' to 'stopped'
Notice: /Stage[main]/Mig::Agent::Daemon/Exec[kill mig]/returns: executed successfully
Info: /Stage[main]/Mig::Agent::Daemon/Exec[kill mig]: Scheduling refresh of Service[mig-agent]
Notice: /Stage[main]/Mig::Agent::Daemon/Service[mig-agent]: Triggered 'refresh' from 1 events
Notice: /Stage[main]/Puppet::Atboot/Exec[reload systemd]/returns: executed successfully
Notice: Finished catalog run in 17.45 seconds

```

**Comment by davehouse** [comment]
I tested on #571 again, and I saw the same failure at first but I'm guessing it was getting the cached copy (showed the correct hash for the version however). On a second run, the change applied without errors or warnings and when I rebooted I see logs of puppet running in the syslog:
```
Jun 26 06:52:09 t-linux64-ms-571.test.releng.mdc2.mozilla.com systemd[1]: Started ACPI event daemon.
Jun 26 06:52:09 t-linux64-ms-571.test.releng.mdc2.mozilla.com systemd[1]: Started Mozilla InvestiGator Agent.
Jun 26 06:52:09 t-linux64-ms-571.test.releng.mdc2.mozilla.com mig-agent-20180409-0.e8eb90a.prod[2849]: [info] Using ext
ernal conf from "/etc/mig/mig-agent.cfg"
Jun 26 06:52:09 t-linux64-ms-571.test.releng.mdc2.mozilla.com puppet-agent[884]: (/Stage[main]/Mig::Agent::Daemon/Exec[
kill mig]/returns) executed successfully
Jun 26 06:52:09 t-linux64-ms-571.test.releng.mdc2.mozilla.com systemd[1]: Stopping Mozilla InvestiGator Agent...
Jun 26 06:52:09 t-linux64-ms-571.test.releng.mdc2.mozilla.com systemd[1]: Stopped Mozilla InvestiGator Agent.
Jun 26 06:52:09 t-linux64-ms-571.test.releng.mdc2.mozilla.com systemd[1]: Started Mozilla InvestiGator Agent.
Jun 26 06:52:09 t-linux64-ms-571.test.releng.mdc2.mozilla.com puppet-agent[884]: (/Stage[main]/Mig::Agent::Daemon/Servi
ce[mig-agent]) Triggered 'refresh' from 1 events
Jun 26 06:52:09 t-linux64-ms-571.test.releng.mdc2.mozilla.com mig-agent-20180409-0.e8eb90a.prod[2873]: [info] Using ext
ernal conf from "/etc/mig/mig-agent.cfg"
Jun 26 06:52:09 t-linux64-ms-571.test.releng.mdc2.mozilla.com puppet-agent[884]: Finished catalog run in 19.26 seconds
Jun 26 06:52:10 t-linux64-ms-571.test.releng.mdc2.mozilla.com systemd[1]: Started Puppet agent.
Jun 26 06:52:10 t-linux64-ms-571.test.releng.mdc2.mozilla.com systemd[1]: Started X11 service.
Jun 26 06:52:10 t-linux64-ms-571.test.releng.mdc2.mozilla.com systemd[1]: Started Xsession service..
Jun 26 06:52:10 t-linux64-ms-571.test.releng.mdc2.mozilla.com systemd[1]: Starting Change resolution to 1600x1200....
```

**Pull request merged**